lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,239 @@
+
+    Args:
+        *packages (str): Packages to upgrade on system.
+    """
+    if len(packages) == 0:
+        raise ExecutionError("No packages specified.")
+    _dnf("upgrade", *packages)
+
+
+def install(*packages: Union[str, os.PathLike]) -> None:
+    """Install one or more packages.
+
+    Args:
+        *packages (Union[str, os.PathLine]): Packages to install on the system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("install", *packages)
+
+
+def remove(*packages: str) -> None:
+    """Remove one or more packages from the system.
+
+    Args:
+        *packages (str): Packages to remove from system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
+
+
+def purge(*packages: str) -> None:
+    """Purge one or more packages from the system.
+
+    Args:
+        *packages (str): Packages to purge from system.
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
+
+
+def fetch(package: str) -> PackageInfo:
+    """Fetch information about a package.
+
+    Args:
+        package (str): Package to get information about.
+
+    Returns:
+        PackageInfo: Information about package.
+    """
+    try:
+        status, info = _dnf("list", "-q", package).split("\n")[:2]  # Only take top two lines.
+        pkg_name, pkg_version, pkg_repo = info.split()
+        name, arch = pkg_name.rsplit(".", 1)
+        epoch, version, release = re.match(r"(?:(.*):)?(.*)-(.*)", pkg_version).groups()
+        if "Installed" in status:
+            state = _PackageState.INSTALLED
+        elif "Available" in status:
+            state = _PackageState.AVAILABLE
+        else:
+            state = _PackageState.UNKNOWN
+
+        return PackageInfo(
+            {
+                "name": name,
+                "arch": arch,
+                "epoch": epoch,
+                "version": version,
+                "release": release,
+                "repo": pkg_repo[1:] if pkg_repo.startswith("@") else pkg_repo,
+                "state": state,
+            }
+        )
+
+    except ExecutionError:
+        return PackageInfo({"name": package, "state": _PackageState.ABSENT})
+
+
+def add_repo(repo: str) -> None:
+    """Add a new repository to DNF.
+
+    Args:
+        repo (str): URL of new repository to add.
+    """
+    if not fetch("dnf-plugins-core").installed:
+        install("dnf-plugins-core")
+    _dnf("config-manager", "--add-repo", repo)
+
+
+def _dnf(*args: str) -> str:
+    """Execute a DNF command.
+
+    Args:
+        *args (str): Arguments to pass to `dnf` executable.
+
+    Raises:
+        ExecutionError: Raised if DNF command execution fails.
+
+    Returns:
+        str: Captured stdout of executed DNF command.
+    """
+    if not installed():
+        raise ExecutionError(f"dnf not found on PATH {os.getenv('PATH')}")
```

Comment:
Should we just let this fail? `subprocess.run` will raise `FileNotFoundError`, so if we did want to catch that explicitly and re-raise `ExecutionError` we could add another `except FileNotFoundError` case below. "It's easier to ask for forgiveness than permission." :-)

Reply:
Let me try this :test_tube: - all I really want to do is ensure that dnf commands are not executed if dnf is not installed on the underlying system.

Reply:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8. Now `FileNotFoundError` is caught rather than performing a manual install check.